File: 2097-valid-arrangement-of-pairs/2097-valid-arrangement-of-pairs.js
var validArrangement = function(pairs) {
    const adj = new Map();
    const inDegree = new Map();
    const outDegree = new Map();
    const path = [];

    const euler = (u) => {
        const stack = [u];
        while (stack.length > 0) {
            const node = stack[stack.length - 1];
            if (!adj.has(node) || adj.get(node).length === 0) {
                path.push(stack.pop());
            } else {
                const child = adj.get(node).pop();
                stack.push(child);
            }
        }
    };

    // Build the graph and track in-degrees and out-degrees
    for (const [u, v] of pairs) {
        if (!adj.has(u)) adj.set(u, []);
        adj.get(u).push(v);
        outDegree.set(u, (outDegree.get(u) || 0) + 1);
        inDegree.set(v, (inDegree.get(v) || 0) + 1);
    }

    // Find the start node
    let start = pairs[0][0];
    for (const [node, outDeg] of outDegree) {
        const inDeg = inDegree.get(node) || 0;
        if (outDeg - inDeg === 1) {
            start = node;
            break;
        }
    }

    // Perform Eulerian path traversal
    euler(start);

    // Construct the result in reverse
    const result = [];
    for (let i = path.length - 1; i > 0; i--) {
        result.push([path[i], path[i - 1]]);
    }
    return result;
};

File: 2097-valid-arrangement-of-pairs/2097-valid-arrangement-of-pairs.cpp
class Solution {
public:
    unordered_map<int,vector<int>>adj;
    vector<int>path;
    void Euler(int u){
        stack<int>stk;
        stk.push(u);
        while(stk.size()){
            int node = stk.top();
            if(adj[node].empty()){
                path.push_back(node);
                stk.pop();
            }
            else {
                int child = adj[node].back();
                adj[node].pop_back();
                stk.push(child);
            }
        }


    }
    vector<vector<int>> validArrangement(vector<vector<int>>& pairs) {
        unordered_map<int,int>vis;
        for(auto it : pairs){
            vis[it[0]]++;
            vis[it[1]]--;
            adj[it[0]].push_back(it[1]);
        }
        int start = pairs[0][0];
        for(auto it : vis)
            if(it.second == 1)start = it.first;
        Euler(start);
        vector<vector<int>>ans;
        // 1 5 4 9 11
        for(int i = path.size()-2 ; i >=0 ;i--){
            ans.push_back({path[i+1],path[i]});
        }
        return ans;

    }
};